db/import_listener.go

```diff
@@ -0,0 +1,133 @@
+}
+
+// StartImportFeed starts an import DCP feed.  Always starts the feed based on previous checkpoints (Backfill:FeedResume).
+// Writes DCP stats into the StatKeyImportDcpStats map
+func (il *importListener) StartImportFeed(bucket base.Bucket, dbStats *DatabaseStats, dbContext *DatabaseContext) error {
+
+	base.Infof(base.KeyDCP, "Attempting to start import DCP feed...")
+
+	il.bucketName = bucket.GetName()
+	il.database = Database{DatabaseContext: dbContext, user: nil}
+	il.stats = dbStats.statsDatabaseMap
+	feedArgs := sgbucket.FeedArguments{
+		ID:         base.DCPImportFeedID,
+		Backfill:   sgbucket.FeedResume,
+		Terminator: il.terminator,
+	}
+
+	importFeedStatsMap, ok := dbContext.DbStats.statsDatabaseMap.Get(base.StatKeyImportDcpStats).(*expvar.Map)
+	if !ok {
+		base.Infof(base.KeyDCP, "Import feed stats map not initialized - will not be tracked")
+	}
+
+	// Start DCP mutation feed
+	base.Infof(base.KeyDCP, "Starting DCP import feed for bucket: %q ", base.UD(bucket.GetName()))
+
+	return bucket.StartDCPFeed(feedArgs, il.ProcessFeedEvent, importFeedStatsMap)
+}
+
+// ProcessFeedEvent is invoked for each mutate or delete event seen on the server's mutation feed.  It may be
+// executed concurrently for multiple events from different vbuckets.  Filters out
+// internal documents based on key, then checks sync metadata to determine whether document needs to be imported
+func (il *importListener) ProcessFeedEvent(event sgbucket.FeedEvent) (shouldPersistCheckpoint bool) {
+
+	shouldPersistCheckpoint = true
+
+	// Ignore non-mutation/deletion events
+	if event.Opcode != sgbucket.FeedOpMutation && event.Opcode != sgbucket.FeedOpDeletion {
+		return shouldPersistCheckpoint
+	}
+	key := string(event.Key)
+
+	// Ignore internal documents
+	if strings.HasPrefix(key, base.SyncPrefix) {
+		if strings.HasPrefix(key, base.DCPCheckpointPrefix) {
+			shouldPersistCheckpoint = false
+		}
+		return shouldPersistCheckpoint
+	}
+
+	// If this is a delete and there are no xattrs (no existing SG revision), we shouldn't import
+	if event.Opcode == sgbucket.FeedOpDeletion && len(event.Value) == 0 {
+		base.Debugf(base.KeyImport, "Ignoring delete mutation for %s - no existing Sync Gateway metadata.", base.UD(event.Key))
+		return shouldPersistCheckpoint
+	}
+
+	// If this is a binary document we can ignore, but update checkpoint to avoid reprocessing upon restart
+	if event.DataType == base.MemcachedDataTypeRaw {
+		return shouldPersistCheckpoint
+	}
+
+	il.ImportFeedEvent(event)
+	return shouldPersistCheckpoint
+}
+
+func (il *importListener) ImportFeedEvent(event sgbucket.FeedEvent) {
+
+	// Unmarshal the doc metadata to determine if this mutation requires import:
+	syncData, rawBody, rawXattr, err := UnmarshalDocumentSyncDataFromFeed(event.Value, event.DataType, false)
+	if err != nil {
+		base.Debugf(base.KeyCache, "Unable to unmarshal sync metadata for feed document %q.  Will not be included in channel cache.  Error: %v", base.UD(event.Key), err)
```

Comment:
These log messages are channel cache related - but I'm also not entirely sure why we're doing this specific error handling. If the doc has no metadata (like below), we're not attempting to do an import?

Reply:
Agree that we could clarify the expectations here (that UnmarshalDocumentSyncDataFromFeed returns no error if metadata isn't found, and that ErrEmptyMetadata represents a case where metadata is expected but hits an error during parsing.

Will also swap out KeyCache for KeyImport.